### PR TITLE
Search using hostname when instance name lookup fails

### LINKED ISSUE #61774: Kubelet cannot register nodes since hostnames are used as node names on OpenStack
**Is this a BUG REPORT**:
/kind bug
/sig openstack

**What happened**:
With #58502 an issue has been introduced to the openstack cloudprovider. This leads to a problem when the kubelet tries to get the `ExternalID` of the node.

https://github.com/kubernetes/kubernetes/blob/e3ac40264fa34bac9e4d1c95e56d1d6203554511/pkg/cloudprovider/providers/openstack/openstack.go#L419-L421

I don't see a way how you can do a server list query with gophercloud by providing the FQDN of an instance.

**Environment**:
- Kubernetes version: 1.10
- Cloud provider or hardware configuration: openstack

### PR DESCRIPTION
**Pull request by dims**
**What this PR does / why we need it**:

We have now switching entirely to using the hostname from the metadata
service or config drive. However we were still searching for the VM
using the server name and not the host name. In this commit, we use
the `hostname=xyz` instead of `name=xyz` which ensures that we make use
of the search using the sanitized name.

https://developer.openstack.org/api-ref/compute/#list-servers
(see hostname (Optional))

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #61774

**Special notes for your reviewer**:

**Release note**:

```release-note
NONE
```

### CODE REVIEW

**Comment by dims** [comment]
cc @voelzmo @BugRoger please take a look

File: pkg/cloudprovider/providers/openstack/openstack.go [hunk 1]
```diff
@@ -416,10 +416,40 @@ func foreachServer(client *gophercloud.ServiceClient, opts servers.ListOptsBuild
...
+	if err == nil {
+		glog.V(5).Infof("Found server : %v/%v", srv.ID, srv.Name)
+		return srv, err
+	}
+	glog.V(5).Infof("Looking up server by host name : %v", name)
+	return getServerByHostName(client, name)
+}
+
+type listOpts struct {
+	Hostname string `q:"hostname"`
```
Comment (voelzmo) [inline pkg/cloudprovider/providers/openstack/openstack.go]: I'm not sure if searching for a hostname works like this? I don't see `hostname` in https://godoc.org/github.com/rackspace/gophercloud/openstack/compute/v2/servers#ListOpts. The nova API says
> This parameter is only valid when specified by administrators. If non-admin users specify this parameter, it is ignored.

See https://developer.openstack.org/api-ref/compute/#list-server-request

So I don't think we can use this filter. It shouldn't be necessary or even recommended to provide an admin user for installing k8s on OpenStack

**Review by voelzmo — CHANGES_REQUESTED** [review]

**Pull request closed without merging**